debian/patches/CVE-2021-3429.patch

```diff
@@ -0,0 +1,141 @@
+             with open(conpath, 'wb') as wfh:
+                 wfh.write(text)
+                 wfh.flush()
+-        else:
++        elif fallback_to_stdout:
+             # A container may lack /dev/console (arguably a container bug).  If
+             # it does not exist, then write output to stdout.  this will result
+             # in duplicate stderr and stdout messages if stderr was True.
+@@ -478,6 +478,26 @@
+     if not o_err:
+         o_err = sys.stderr
+ 
++    # pylint: disable=subprocess-popen-preexec-fn
++    def set_subprocess_umask_and_gid():
++        """Reconfigure umask and group ID to create output files securely.
++
++        This is passed to subprocess.Popen as preexec_fn, so it is executed in
++        the context of the newly-created process.  It:
++
++        * sets the umask of the process so created files aren't world-readable
++        * if an adm group exists in the system, sets that as the process' GID
++          (so that the created file(s) are owned by root:adm)
++        """
++        os.umask(0o037)
++        try:
++            group_id = grp.getgrnam("adm").gr_gid
++        except KeyError:
++            # No adm group, don't set a group
++            pass
++        else:
++            os.setgid(group_id)
++
+     if outfmt:
+         LOG.debug("Redirecting %s to %s", o_out, outfmt)
+         (mode, arg) = outfmt.split(" ", 1)
+@@ -487,7 +507,12 @@
+                 owith = "wb"
+             new_fp = open(arg, owith)
+         elif mode == "|":
+-            proc = subprocess.Popen(arg, shell=True, stdin=subprocess.PIPE)
++            proc = subprocess.Popen(
++                arg,
++                shell=True,
++                stdin=subprocess.PIPE,
++                preexec_fn=set_subprocess_umask_and_gid,
++            )
+             new_fp = proc.stdin  # pylint: disable=E1101
+         else:
+             raise TypeError("Invalid type for output format: %s" % outfmt)
+@@ -509,7 +534,12 @@
+                 owith = "wb"
+             new_fp = open(arg, owith)
+         elif mode == "|":
+-            proc = subprocess.Popen(arg, shell=True, stdin=subprocess.PIPE)
++            proc = subprocess.Popen(
++                arg,
++                shell=True,
++                stdin=subprocess.PIPE,
++                preexec_fn=set_subprocess_umask_and_gid,
++            )
+             new_fp = proc.stdin  # pylint: disable=E1101
+         else:
+             raise TypeError("Invalid type for error format: %s" % errfmt)
+--- a/tests/unittests/test_util.py
++++ b/tests/unittests/test_util.py
+@@ -307,4 +307,10 @@
+         expected = ('none', 'tmpfs', '/run/lock')
+         self.assertEqual(expected, util.parse_mount_info('/run/lock', lines))
+ 
++class TestMultiLog(helpers.FilesystemMockingTestCase):
++    def test_logs_dont_go_to_stdout_if_fallback_to_stdout_is_false(self):
++        util.multi_log('something', fallback_to_stdout=False)
++        self.assertEqual('', self.stdout.getvalue())
++
++
```

Comment:
This is where the issue is atm 😞 

The version in Debian Jessie (0.7.6~bzr976) is a bit too old and doesn't have `tests/integration_tests/*` or `cloudinit/config/tests/*` or `cloudinit/tests/*` either. It just has `tests/unittests/`, so I've just backported this bit from https://github.com/canonical/cloud-init/commit/b794d426b9ab43ea9d6371477466070d86e10668 to see how it goes and it did not work alright! :/

If I build, the tests fail:
```
Running tests...
..................................................................................................................................................................................................................................................................................................................................somethingE.....
======================================================================
ERROR: test_logs_dont_go_to_stdout_if_fallback_to_stdout_is_false (tests.unittests.test_util.TestMultiLog)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/mocker.py", line 146, in test_method_wrapper
    result = test_method()
  File "/<<PKGBUILDDIR>>/tests/unittests/test_util.py", line 313, in test_logs_dont_go_to_stdout_if_fallback_to_stdout_is_false
    self.assertEqual('', self.stdout.getvalue())
AttributeError: 'TestMultiLog' object has no attribute 'stdout'
```

This clearly looks like I am missing something or doing the backport incorrectly. But I couldn't figure it out and would like to get another pair of eyes on this.